From 43b31be30f7ec6f5212d8ca4282e79f97cc0351a Mon Sep 17 00:00:00 2001
Message-ID: <43b31be30f7ec6f5212d8ca4282e79f97cc0351a.1776946505.git.bertrand.marquis@arm.com>
In-Reply-To: <55d56a174a0f04520d4a1558630bb0b739074b13.1776946505.git.bertrand.marquis@arm.com>
References: <55d56a174a0f04520d4a1558630bb0b739074b13.1776946505.git.bertrand.marquis@arm.com>
From: Bertrand Marquis <bertrand.marquis@arm.com>
Date: Thu, 16 Apr 2026 19:30:46 +0200
Subject: [RFC PATCH 15/28] virtio-msg: derive DMA address spaces from the
 owning proxy

Resolve the owning VirtIOMSGProxy from child devices before selecting
the DMA AddressSpace.

This lets shared virtio-msg helpers derive DMA state from the proxy
even when the caller only has a child device pointer.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>
---
 hw/virtio/virtio-msg.c | 24 ++++++++++++++++++++----
 1 file changed, 20 insertions(+), 4 deletions(-)

diff --git a/hw/virtio/virtio-msg.c b/hw/virtio/virtio-msg.c
index e77bd47e0f74..3ce05c00fa72 100644
--- a/hw/virtio/virtio-msg.c
+++ b/hw/virtio/virtio-msg.c
@@ -860,13 +860,29 @@ static void virtio_msg_pre_plugged(DeviceState *d, Error **errp)
     virtio_add_feature(&vdev->host_features, VIRTIO_F_VERSION_1);
 }
 
+static VirtIOMSGProxy *virtio_msg_resolve_proxy(DeviceState *d)
+{
+    VirtIOMSGProxy *s = NULL;
+
+    if (object_dynamic_cast(OBJECT(d), TYPE_VIRTIO_MSG)) {
+        s = VIRTIO_MSG(d);
+    } else if (object_dynamic_cast(OBJECT(d), TYPE_VIRTIO_MSG_DEV)) {
+        VirtIOMSGDev *mdev = VIRTIO_MSG_DEV(d);
+        s = mdev->proxy;
+    }
+
+    return s;
+}
+
 static AddressSpace *virtio_msg_get_dma_as(DeviceState *d)
 {
-    VirtIOMSGProxy *s = VIRTIO_MSG(d);
-    AddressSpace *as;
+    VirtIOMSGProxy *s = virtio_msg_resolve_proxy(d);
+
+    if (!s) {
+        return &address_space_memory;
+    }
 
-    as = virtio_msg_bus_get_remote_as(&s->msg_bus);
-    return as;
+    return virtio_msg_bus_get_remote_as(&s->msg_bus);
 }
 
 static int virtio_msg_query_nvectors(DeviceState *d)
-- 
2.53.0

